### PR TITLE
Support log trace injection for log4j 1.x and log4j 2.x when used without Slf4j

### PR DESCRIPTION
**Pull request by labbati**
Previously we only supported log correlation when the java application was using Slf4j.

This PR add support for trace <-> log correlation outside Slf4j if the app is using log4j 1.x or logj 2.x.

In addition to that we only defined a mandatory test suite that a logging library integration must satisfy in order to properly support trace correlation.

### CODE REVIEW

File: dd-java-agent/instrumentation/slf4j-mdc/src/test/groovy/Slf4jMDCTest.groovy [hunk 1]
```diff
@@ -1,95 +1,15 @@
...
-    ConfigUtils.updateConfig {
-      System.setProperty("dd.logs.injection", "true")
-    }
+  @Override
+  def put(String key, Object value) {
+    return MDC.put(key, value as String)
   }
 
-  def "mdc shows trace and span ids for active scope"() {
-    when:
```
Comment (labbati) [inline dd-java-agent/instrumentation/slf4j-mdc/src/test/groovy/Slf4jMDCTest.groovy]: Tests moved to the abstract  class: https://github.com/DataDog/dd-trace-java/pull/894/files#diff-076e0aa63abc1c8a6bf27f11c7ceb6dbR14

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentInstaller.java [hunk 2]
```diff
@@ -128,6 +128,7 @@ public static ResettableClassFileTransformer installBytebuddyAgent(
             .or(nameStartsWith("com.jinspired."))
             .or(nameStartsWith("org.jinspired."))
             .or(nameStartsWith("org.apache.log4j."))
+            .and(not(named("org.apache.log4j.MDC")))
```
Comment (mar-kolya) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentInstaller.java]: This seems wrong - `and` should be inside `or`, shouldn't it?

**Comment by mar-kolya** [comment]
You may want to look into why tests failing there - looks like jacoco needs some tuning

File: dd-java-agent/instrumentation/log4j1/src/main/java/datadog/trace/instrumentation/log4j1/Log4j1MDCInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  @Override
+  public ElementMatcher<? super TypeDescription> typeMatcher() {
+    return named("org.apache.log4j.MDC");
+  }
+
+  @Override
+  public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+    return singletonMap(isConstructor(), MDCContextAdvice.class.getName());
```
Comment (mar-kolya) [inline dd-java-agent/instrumentation/log4j1/src/main/java/datadog/trace/instrumentation/log4j1/Log4j1MDCInstrumentation.java]: `org.apache.log4j.MDC` seems to be class with only static methods. It seems slightly awkward to instrument it's constructor. Are you sure it is being called at all?

Comment (mar-kolya) [reply dd-java-agent/instrumentation/log4j1/src/main/java/datadog/trace/instrumentation/log4j1/Log4j1MDCInstrumentation.java]: Ah, nm, I was looking at only one implementation out of a few... using constructor as advice hook still seems odd.

File: dd-java-agent/instrumentation/log4j2/log4j2.gradle [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    versions = '(,)'
+  }
+}
+
+dependencies {
+  compile project(':dd-trace-api')
+  compile project(':dd-java-agent:agent-tooling')
+
+  testCompile group: 'org.apache.logging.log4j', name: 'log4j-core', version: log4jVersion
+  testCompile group: 'org.apache.logging.log4j', name: 'log4j-api', version: log4jVersion
```
Comment (mar-kolya) [inline dd-java-agent/instrumentation/log4j2/log4j2.gradle]: nitpic: you may want to group `testCompile` clauses together and put them after non-test dependencies.

File: dd-java-agent/instrumentation/log4j2/src/main/java/datadog/trace/instrumentation/log4j2/ThreadContextInstrumentation.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import java.lang.reflect.Method;
+import java.util.Map;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+@AutoService(Instrumenter.class)
+public class ThreadContextInstrumentation extends Instrumenter.Default {
+  public static final String MDC_INSTRUMENTATION_NAME = "log4j-thread-context";
```
Comment (mar-kolya) [inline dd-java-agent/instrumentation/log4j2/src/main/java/datadog/trace/instrumentation/log4j2/ThreadContextInstrumentation.java]: Should naming simply be `log4j1` and `log4j2` so make things less confusing?

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentInstaller.java (hunk 2 above)
Comment (labbati) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentInstaller.java]: Good catch, at the end of the day it worked but it was very wrong 😄

File: dd-java-agent/instrumentation/log4j1/src/main/java/datadog/trace/instrumentation/log4j1/Log4j1MDCInstrumentation.java (hunk 3 above)
Comment (labbati) [reply dd-java-agent/instrumentation/log4j1/src/main/java/datadog/trace/instrumentation/log4j1/Log4j1MDCInstrumentation.java]: It took a while to find out the source code corresponding to the version being downloaded by gradle. Here is the link for future reference: http://svn.apache.org/viewvc/logging/log4j/trunk/src/main/java/org/apache/log4j/MDC.java?view=markup

File: dd-java-agent/instrumentation/log4j2/src/main/java/datadog/trace/instrumentation/log4j2/ThreadContextInstrumentation.java (hunk 5 above)
Comment (labbati) [reply dd-java-agent/instrumentation/log4j2/src/main/java/datadog/trace/instrumentation/log4j2/ThreadContextInstrumentation.java]: That is a good idea. Alternatively we could name it `mdc` in the same way it happens for the Slf4j instrumentation. WDYT?

Comment (mar-kolya) [reply dd-java-agent/instrumentation/log4j2/src/main/java/datadog/trace/instrumentation/log4j2/ThreadContextInstrumentation.java]: I do not have a strong preference

**Comment by labbati** [comment]
@mar-kolya might I have your final approval to merge?

File: dd-java-agent/instrumentation/log4j1/src/main/java/datadog/trace/instrumentation/log4j1/Log4j1MDCInstrumentation.java (hunk 3 above)
Comment (mar-kolya) [reply dd-java-agent/instrumentation/log4j1/src/main/java/datadog/trace/instrumentation/log4j1/Log4j1MDCInstrumentation.java]: You do not have to change this now, or at all - but I think `isTypeInitializer` might have been better matcher here.

**Pull request merged**